File: Hashing/L2_Implementation_of_Chaining.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;
struct MyHash
{
    int BUCKET;
    list<int> *table;
    MyHash(int b)
    {
        BUCKET=b;
        table=new list<int> [b];
    }
    void insert(int key)
    {
        int i=key%BUCKET;
        table[i].push_back(key);
    }
    void remove(int key)
    {
        int i = key%BUCKET;
        table[i].remove(key);
    }
    bool search(int key)
    {
        int i = key%BUCKET;
        for(auto x:table[i])
        {
            if(x==key)
                return true;
        }
        return false;
    }
};
int main(){
    // struct MyHash mh(7);
    //     or   ---> Any of both can use
    // MyHash mh(7);
    MyHash mh(7);
    mh.insert(10);
    mh.insert(20);
    mh.insert(15);
    mh.insert(7);
    cout<<mh.search(10)<<endl;
    mh.remove(20);
    cout<<mh.search(20)<<endl;
    
    return 0;
}

File: code_java/self pased DSA/Hashing/L2_Implementation_of_Chaining.java
/*
Problem: Implement a hash table using chaining for collision resolution.
Approach:
- Use an array of linked lists (ArrayList of LinkedLists) to store keys. Insert, search, and remove use chaining.
*/
import java.util.*;

public class L2_Implementation_of_Chaining {
    static class MyHash {
        int BUCKET;
        List<LinkedList<Integer>> table;
        MyHash(int b) {
            BUCKET = b;
            table = new ArrayList<>(b);
            for (int i = 0; i < b; i++) table.add(new LinkedList<>());
        }
        void insert(int key) {
            int i = key % BUCKET;
            table.get(i).add(key);
        }
        void remove(int key) {
            int i = key % BUCKET;
            table.get(i).remove((Integer) key);
        }
        boolean search(int key) {
            int i = key % BUCKET;
            return table.get(i).contains(key);
        }
    }
    public static void main(String[] args) {
        MyHash mh = new MyHash(7);
        mh.insert(10);
        mh.insert(20);
        mh.insert(15);
        mh.insert(7);
        System.out.println(mh.search(10));
        mh.remove(20);
        System.out.println(mh.search(20));
    }
} 